x-pack/plugins/stack_connectors/public/components/text_area_with_autocomplete.tsx

```diff
@@ -0,0 +1,328 @@
+import { filterSuggestions } from '../lib/filter_suggestions_for_autocomplete';
+import { templateActionVariable } from '../lib/template_action_variable';
+
+export interface TextAreaWithAutocompleteProps {
+  editAction: (property: string, value: any, index: number) => void;
+  errors?: string[];
+  index: number;
+  inputTargetValue?: string;
+  isDisabled?: boolean;
+  label: string;
+  messageVariables?: ActionVariable[];
+  paramsProperty: string;
+}
+const selectableListProps = { className: 'euiSelectableMsgAutoComplete' };
+
+export const TextAreaWithAutocomplete: React.FunctionComponent<TextAreaWithAutocompleteProps> = ({
+  editAction,
+  errors,
+  index,
+  inputTargetValue,
+  isDisabled = false,
+  label,
+  messageVariables,
+  paramsProperty,
+}) => {
+  const { euiTheme } = useEuiTheme();
+  const backgroundColor = useEuiBackgroundColor('plain');
+
+  const textAreaRef = React.useRef<HTMLTextAreaElement | null>(null);
+  const selectableRef = React.useRef<EuiSelectable | null>(null);
+
+  const [matches, setMatches] = useState<string[]>([]);
+  const [popupPosition, setPopupPosition] = useState({ top: 0, left: 0, height: 0, width: 0 });
+  const [isListOpen, setListOpen] = useState(false);
+  const [selectableHasFocus, setSelectableHasFocus] = useState(false);
+  const [searchWord, setSearchWord] = useState<string | null>(null);
+
+  const optionsToShow: EuiSelectableOption[] = useMemo(() => {
+    return matches?.map((variable) => ({
+      label: variable,
+      data: {
+        description: variable,
+      },
+      'data-test-subj': `${variable}-selectableOption`,
+    }));
+  }, [matches]);
+
+  const onOptionPick = useCallback(
+    (newOptions: EuiSelectableOption[]) => {
+      if (!textAreaRef.current) return;
+      const { value, selectionStart, scrollTop } = textAreaRef.current;
+      const lastSpaceIndex = value.slice(0, selectionStart).lastIndexOf(' ');
+      const lastOpenDoubleCurlyBracketsIndex = value.slice(0, selectionStart).lastIndexOf('{{');
+      const currentWordStartIndex = Math.max(lastSpaceIndex, lastOpenDoubleCurlyBracketsIndex);
+
+      const checkedElement = newOptions.find(({ checked }) => checked === 'on');
+      if (checkedElement) {
+        const newInputText =
+          value.slice(0, currentWordStartIndex) +
+          '{{' +
+          checkedElement.label +
+          '}}' +
+          value.slice(selectionStart);
+
+        editAction(paramsProperty, newInputText.trim(), index);
+        setMatches([]);
+        textAreaRef.current.focus();
+        // We use setTimeout here, because editAction is async function and we need to wait before it executes
+        setTimeout(() => {
+          if (textAreaRef.current) {
+            textAreaRef.current.selectionStart =
+              currentWordStartIndex + checkedElement.label.length + 4;
+            textAreaRef.current.selectionEnd = textAreaRef.current.selectionStart;
+            textAreaRef.current.scrollTop = scrollTop;
```

Comment:
I've put comment here: We use setTimeout here, because editAction is async function and we need to wait before it executes.
We need to wait here before `editAction(paramsProperty, newInputText.trim(), index)` will be executed, because this command put cursor at the end of the input. 
And after that we will recalculate cursor position:
```
if (textAreaRef.current) {
            textAreaRef.current.selectionStart =
              currentWordStartIndex + checkedElement.label.length + 4;
            textAreaRef.current.selectionEnd = textAreaRef.current.selectionStart;
            textAreaRef.current.scrollTop = scrollTop;
          }
```
Without setTimeout first will be executed code, which now is in setTimeout, but `editAction` still will be running and put cursor in wrong position in the end.

Maybe I can do it in the better way. But I need hints.

Reply:
@XavierM you wanted to write something here about setTimeout.

Reply:
Using setTimeout is not ideal but it's not worth spending too much time fixing either.